plugins/openssl/openssl.cpp

```diff
@@ -0,0 +1,950 @@
+        }
+        else
+            failOpenSSLError("adding new object to cache");
+
+        return newObj;
+    };
+
+    void printStatistics() {DBGLOG("OPENSSL %s CACHE STATS: HITS = %d, MISSES = %d", cacheName.c_str(), hits, misses);};
+
+    void init()
+    {
+        setCacheName();
+        hits = 0;
+        misses = 0;
+    };
+
+    void clear() {cache.clear();};
+
+private:
+    int hits;
+    int misses;
+    std::string cacheName;
+    std::list<std::tuple<std::string, const T *>> cache;
+
+    void setCacheName();
+    const T * getObjectByName(const char * name);
+};
+
+template <>
+void OpenSSLCache<EVP_CIPHER>::setCacheName() { cacheName = "CIPHER"; }
+
+template <>
+void OpenSSLCache<EVP_MD>::setCacheName() { cacheName = "DIGEST"; }
+
+template <>
+const EVP_CIPHER * OpenSSLCache<EVP_CIPHER>::getObjectByName(const char * name) { return EVP_get_cipherbyname(name); }
+
+template <>
+const EVP_MD * OpenSSLCache<EVP_MD>::getObjectByName(const char * name) { return EVP_get_digestbyname(name); }
+
+// PEM Public/Private keys require parsing from a string
+// Store the hash of the original string and parsed key
+class PKeyCache
+{
+public:
+    void init()
+    {
+        hits = 0;
+        misses = 0;
+    };
+
+    EVP_PKEY * checkCache(size32_t keyLen, const char * key, size32_t passphraseLen, const void * passphrase)
+    {
+        for (auto& c : cache)
+        {
+            if (hashc(reinterpret_cast<const byte *>(passphrase), passphraseLen, hashc(reinterpret_cast<const byte *>(key), keyLen, 0)) == std::get<0>(c))
+            {
+                hits++;
+                return std::get<1>(c);
+            }
+        }
+
+        misses++;
+
+        BIO * bio = BIO_new_mem_buf(key, keyLen);
+        if (!bio)
+            failOpenSSLError("creating buffer for EVP_PKEY");
+
+        EVP_PKEY * pkey;
+        if (isPublicKey(keyLen, key))
+            pkey = PEM_read_bio_PUBKEY(bio, nullptr, nullptr, nullptr);
+        else
+        {
+            MemoryBuffer passphraseMB;
+            passphraseMB.setBuffer(passphraseLen, (void *)passphrase);
+            pkey = PEM_read_bio_PrivateKey(bio, nullptr, passphraseCB, static_cast<void *>(&passphraseMB));
+        }
+        BIO_free(bio);
+
+        if (pkey)
+        {
+            cache.emplace_front(hashc(reinterpret_cast<const byte *>(passphrase), passphraseLen, hashc(reinterpret_cast<const byte *>(key), keyLen, 0)), pkey);
+            if (cache.size() > OPENSSL_MAX_CACHE_SIZE)
+            {
+                EVP_PKEY_free(std::get<1>(cache.back()));
+                cache.pop_back();
+            }
+        }
+        else
+            failOpenSSLError("loading a pkey");
+
+        return pkey;
+    };
+
+    void clear()
+    {
+        for (auto& c : cache)
+            EVP_PKEY_free(std::get<1>(c));
+        cache.clear();
+    };
+
+    void printStatistics() {DBGLOG("OPENSSL PKEY CACHE STATS: HITS = %d, MISSES = %d", hits, misses);};
+
+private:
+    int hits;
```

Comment:
minor: unsigned would be better since they can never be negative, and avoid overflow issues if they wrap.